### PR TITLE
Feature/123 project 1

### PR DESCRIPTION
**Pull request by MidhatAsif**
- CRUD operations for Product model
- Test cases for Project Controller + Project Model

### CODE REVIEW

**Comment by MidhatAsif** [comment]
Cycle-0

File: spec/controllers/projects_controller_spec.rb [hunk 1]
```diff
@@ -0,0 +1,161 @@
...
+
+    context 'valid attributes' do
+      it 'locates the requested @project' do
+        put :update, id: @project, project: FactoryGirl.attributes_for(:project)
+        expect(assigns(:project)).to eq(@project)
+      end
+
+      it "changes @project's attributes" do
+        put :update, id: @project, project: FactoryGirl.attributes_for(
+          :project, title: 'Edit Test Proj',
```
Comment (abubakarazeem) [inline spec/controllers/projects_controller_spec.rb]: store title and description in a variable or use let here.. DRY

File: spec/controllers/projects_controller_spec.rb [hunk 2]
```diff
@@ -0,0 +1,161 @@
...
+    it 'renders the :new template' do
+      get :new
+      expect(response).to render_template(:new)
+    end
+  end
+
+  describe 'POST #create' do
+    context 'with valid attributes' do
+      it 'saves the new project in the database' do
+        expect { post :create, project: FactoryGirl.attributes_for(:project) }
```
Comment (abubakarazeem) [inline spec/controllers/projects_controller_spec.rb]: let(:project_attr) { FactoryGirl.attributes_for(:project) } can be used here

File: spec/controllers/projects_controller_spec.rb [hunk 3]
```diff
@@ -0,0 +1,161 @@
...
+      end
+      it "redirects to that new project's page" do
+        post :create, project: FactoryGirl.attributes_for(:project)
+        expect(response).to redirect_to Project.last
+      end
+    end
+
+    context 'with invalid attributes' do
+      it 'does not save the new project in the database' do
+        expect { post :create, project: FactoryGirl.attributes_for(:invalid_project) }
```
Comment (abubakarazeem) [inline spec/controllers/projects_controller_spec.rb]: same as above.. let can be used.. DRY

File: spec/controllers/projects_controller_spec.rb [hunk 4]
```diff
@@ -0,0 +1,161 @@
...
+      it 'locates the requested @project' do
+        put :update, id: @project,
+                     project: FactoryGirl.attributes_for(:invalid_project)
+        expect(assigns(:project)).to eq(@project)
+      end
+
+      it "does not change @project's attributes" do
+        put :update, id: @project,
+                     project: FactoryGirl.attributes_for(
+                       :project, title: 'Edit Test Proj', description: nil
```
Comment (abubakarazeem) [inline spec/controllers/projects_controller_spec.rb]: variable to store title..

File: spec/controllers/projects_controller_spec.rb [hunk 5]
```diff
@@ -0,0 +1,161 @@
+require 'rails_helper'
+require 'spec_helper'
+
+RSpec.describe ProjectsController, type: :controller do
```
Comment (abubakarazeem) [inline spec/controllers/projects_controller_spec.rb]: 'should' is missing in most of the test descriptions =P

File: spec/models/project_spec.rb [hunk 6]
```diff
@@ -0,0 +1,15 @@
+require 'rails_helper'
+
+RSpec.describe Project, type: :model do
```
Comment (abubakarazeem) [inline spec/models/project_spec.rb]: i think 'should' word should be included in test description =P

File: spec/controllers/projects_controller_spec.rb (hunk 5 above)
Comment (MidhatAsif) [reply spec/controllers/projects_controller_spec.rb]: "should" is the deprecated syntax. :p

File: spec/models/project_spec.rb (hunk 6 above)
Comment (MidhatAsif) [reply spec/models/project_spec.rb]: I warned me that i was using old syntax and prompted me to use the new "expect" syntax :p

File: app/models/project.rb [hunk 7]
```diff
@@ -1,4 +1,7 @@
 class Project < ActiveRecord::Base
+  validates :title, presence: true
+  validates :description, presence: true
```
[the remark this answers is not in the document]
Comment (MidhatAsif) [reply app/models/project.rb]: Done + wrote it's test cases too

File: spec/controllers/projects_controller_spec.rb (hunk 4 above)
Comment (MidhatAsif) [reply spec/controllers/projects_controller_spec.rb]: Done for description too.

File: spec/controllers/projects_controller_spec.rb (hunk 5 above)
Comment (MidhatAsif) [reply spec/controllers/projects_controller_spec.rb]: places should in statements :p

File: spec/models/project_spec.rb (hunk 6 above)
Comment (MidhatAsif) [reply spec/models/project_spec.rb]: placed "should" in descriptions

**Comment by MidhatAsif** [comment]
Pushed changes for cycle-1 review.

**Pull request merged**